graphql/src/execution/execution.rs

```diff
@@ -15,9 +15,57 @@ use crate::prelude::*;
 use crate::query::ast as qast;
 use crate::schema::ast as sast;
 use crate::values::coercion;
+use lru_time_cache::LruCache;
+use std::sync::Mutex;
 
 lazy_static! {
     static ref NO_PREFETCH: bool = std::env::var_os("GRAPH_GRAPHQL_NO_PREFETCH").is_some();
+
+    // Comma separated subgraph ids to cache queries for.
+    static ref CACHED_SUBGRAPH_IDS: Vec<String> = {
+        std::env::var("GRAPH_CACHED_SUBGRAPH_IDS")
+        .unwrap_or_default()
+        .split(',')
+        .map(|s| s.to_owned())
+        .collect()
+    };
+
+    // Cache expiry time, 1 minute by default.
+    static ref CACHE_EXPIRY: Duration = {
+        std::env::var("GRAPH_CACHE_EXPIRY_SECS")
+        .unwrap_or("60".to_string())
+        .parse::<u64>()
+        .map(|s| Duration::from_secs(s))
+        .unwrap()
+    };
+
+    // This cache might serve stale data.
+    static ref QUERY_CACHE: Mutex<LruCache<[u8; 32], BTreeMap<String, q::Value>>>
+                                = Mutex::new(LruCache::with_expiry_duration(*CACHE_EXPIRY));
+}
+
+// The key is: subgraph id + selection set + variables + fragment definitions
+fn cache_key(ctx: &ExecutionContext<impl Resolver>, selection_set: &q::SelectionSet) -> [u8; 32] {
+    let mut hasher = blake3::Hasher::new();
+    // subgraph id
+    hasher.update(ctx.query.schema.id.to_string().as_bytes());
+
+    // selection set
+    hasher.update(selection_set.to_string().as_bytes());
+
+    // variables
+    for (key, value) in &ctx.query.variables {
+        hasher.update(key.as_bytes());
```

Comment:
This is not collision-resistant.

Consider, for example, a query variable with the key "KEY", and the value "VALUE". This hashes the same as a query variable with the key "KEYV" and the value "ALUE". In both cases, all that the hasher sees is the string "KEYVALUE", so the hash is the same.

We worked through all this and more in the `StableHash` crate. I would strongly consider using that which will help to get it right automatically. I can help with that effort if necessary.